server/aesdsocket.c

```diff
@@ -67,227 +110,348 @@ int deamon_create()
+	openlog(NULL , 0 , LOG_USER);
+	printf("Welcome to Socket \n");
+	memset(&hints, 0, sizeof hints); 
+	hints.ai_family = AF_UNSPEC;    
+	hints.ai_socktype = SOCK_STREAM; 
+	hints.ai_flags = AI_PASSIVE;    
+	status =  getaddrinfo(NULL,  "9000" , &hints, &result);
+        if (status != 0)
+        {
+      		syslog(LOG_ERR,"Getaddrinfo error");  	
+      		exit(-1);
+        }
+        
+
+        for(temp_ptr = result; temp_ptr != NULL; temp_ptr = temp_ptr->ai_next)
+        { 
+        	socket_fd= socket(result->ai_family, result->ai_socktype, 0);
+        	if (socket_fd == -1)
+        	{
+        		syslog(LOG_ERR,"Socket Creation error");  	
+      			exit(-1);
+        
+       	}
+       	status = setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &(int){1}, sizeof(int));
+		status= bind(socket_fd, result->ai_addr, result->ai_addrlen);
+		if (status == 0)
+        	{
+        		syslog(LOG_DEBUG,"Bind is successful");  	
+      			break; 
+       	}
+       }
+        
+       freeaddrinfo(result);
+       if(temp_ptr == NULL)
+       {
+       	syslog(LOG_DEBUG,"ERROR: bind() fail");
+       	exit(-1);
+       
+       }
+	if( argc > 1 && !strcmp("-d", argv[1]) ) 
+	{
+		syslog(LOG_INFO, "Run as daemon");
+		daemon_create();
+	}
+	status= listen(socket_fd, BACKLOGS);
+       if (status == -1)
+       {
+        	syslog(LOG_ERR,"Error in listen");  	
+      		exit(-1);
+        
+        }
+        
+       	open_fd = creat("/var/tmp/aesdsocketdata", 0644);
+        	if (open_fd == -1)
+        	{
+        		syslog(LOG_ERR,"File cannot be created");
+        		exit(-1);        	
+        	}
+      
+      	status = setitimer (ITIMER_REAL, &delay, NULL);
+	if(status)
+	{
+		syslog(LOG_ERR, "Error in timer set");
+		exit (-1);
+	}
+	
+	status = pthread_mutex_init(&mutex1, NULL);
+	if (status != 0)
+	{
+		syslog(LOG_ERR, "Mutex Initialization error");
+		exit (-1);
+	
+	}
+        while (1)
+        {
+
+        	host_address_size = sizeof(struct sockaddr);
+        	accept_fd= accept(socket_fd, (struct sockaddr *)&host_address, &host_address_size);
+        	if (accept_fd == -1)
+        	{
+        		syslog(LOG_ERR,"Accept comamnd error");
+        		exit(-1);
+        	}
+        	syslog(LOG_DEBUG, "Accepted connection from %s\n", inet_ntoa(host_address.sin_addr)); 
+        	printf("Accepted connection from %s\n", inet_ntoa(host_address.sin_addr));
+        	add_thread = (slist_data_t *) malloc(sizeof(slist_data_t));  
+        	SLIST_INSERT_HEAD(&head, add_thread, entries);     
+        	add_thread->thread_data_args.mutex = &mutex1;
+        	add_thread->thread_data_args.accept_socket_fd = accept_fd;
+        	add_thread->thread_data_args.thread_complete_success = false;
+        	pthread_create(&(add_thread->thread_data_args.thread_id), NULL, thread_func, &(add_thread->thread_data_args));
+        	
+
+        	SLIST_FOREACH(add_thread, &head, entries)
+        	{
+        		pthread_join(add_thread->thread_data_args.thread_id, NULL);
```

Comment:
@Rajs060699 you will only have one active thread if you pthread_join all threads immediatly after creating.  You need to check for thread completion here.